### ISSUE TITLE
Provider v3.56.0 recreated table after adding nullable column

### ISSUE THREAD

**Issue by psvaiter** [hashicorp/terraform-provider-google#8503]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

v0.12.24

### Affected Resource(s)

* google_bigquery_table

### Terraform Configuration Files

I don't have access to this configuration.

### Debug Output

I don't have access to the debug output as it's part of deployment continuous delivery.

### Panic Output

N/A

### Expected Behavior

Add nullable column to table.

### Actual Behavior

Recreated table (losing data). The plus sign at the left of plan output showed the column being added, and that made me think the terraform plan was ok. However after the incident I noticed there's a `forces replacement` comment a the beginning of fields' list.

I replaced the project and dataset names.

```
   # module.gcp_my_dataset.module.events.module.bigquery_table.google_bigquery_table.default[1] must be replaced
 -/+ resource "google_bigquery_table" "default" {
       ~ creation_time       = 1591142718086 -> (known after apply)
         dataset_id          = "dataset_1"
       ~ etag                = "iZ0lNwUSrJHIOCmuNTIk3A==" -> (known after apply)
       ~ expiration_time     = 0 -> (known after apply)
       ~ id                  = "projects/stuff-prd/datasets/dataset_1/tables/loan" -> (known after apply)
         labels              = {
             "env" = "default"
         }
       ~ last_modified_time  = 1613660554887 -> (known after apply)
       ~ location            = "US" -> (known after apply)
       ~ num_bytes           = 70295692884 -> (known after apply)
       ~ num_long_term_bytes = 32523024442 -> (known after apply)
       ~ num_rows            = 117726253 -> (known after apply)
       ~ project             = "stuff-prd" -> (known after apply)
       ~ schema              = jsonencode(
           ~ [ # forces replacement
                 {
                     mode = "REQUIRED"
                     name = "Id"
                     type = "STRING"
                 },
                 {
                     mode = "REQUIRED"
                     name = "CreationDate"
                     type = "TIMESTAMP"
                 },
                 {
                     mode = "NULLABLE"
                     name = "LastUpdateDate"
                     type = "TIMESTAMP"
                 },
                 {
                     mode = "NULLABLE"
                     name = "MaximumMonthlyRetentionAmount"
                     type = "FLOAT"
                 },
               + {
                   + mode = "NULLABLE"
                   + name = "MaximumRetentionAmount"
                   + type = "FLOAT"
                 },
             ]
         )
       ~ self_link           = "https://bigquery.googleapis.com/bigquery/v2/projects/stuff-prd/datasets/dataset_1/tables/loan" -> (known after apply)
         table_id            = "loan"
       ~ type                = "TABLE" -> (known after apply)
 
       ~ time_partitioning {
           ~ expiration_ms            = 0 -> (known after apply)
             field                    = "LastUpdateDate"
           - require_partition_filter = false -> null
             type                     = "DAY"
         }
     }
```

### Steps to Reproduce

1. `terraform plan`
2. `terraform apply`

### Important Factoids

We're using the following provider version:
```terraform
provider "google" {
  project = "stuff-prd"
  region  = "us-east1"
  zone    = "us-east1-b"
  version = "~> 3.56.0"
}
```

### References

The CHANGELOG points to a bug in v3.54.0 that was reverted with a breaking change in v3.55.0 So I thought it was secure to update from v3.53.0 to v3.56.0 as I was interested in the following bugfix:

- bigquery: fixed bug where you could not reorder columns on schema for resource google_bigquery_table (#8321)

However the table was silently recreated using this version of the provider, producing an incident of data loss. Fortunately I could recover data accessing a previous version of the table.

**Comment by edwardmedia** [comment]
@psvaiter Can you share your config and specify how you added the column(s)?

**Comment by rm3195** [comment]
Confirming this issue on v3.57.0 and terraform 0.12.24, it's very unexpected behavior. Luckily we have protectors in place for production tables...
In our case schemas are defined in json files, so simply adding a column to json file forces table recreation.

**Comment by edwardmedia** [comment]
@psvaiter yes, recreation is by design and expected when the schema is altered because it is a [JSON schema](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_table#schema) as a whole. Closing this issue then

**Comment by psvaiter** [comment]
Recreation was not by design in previous versions, so I don't understand how this is the expected behavior. This should even be considered a breaking change as this cause data loss in production tables.

I don't know if I was clear about adding nullable column, which should NOT imply in table recreation.

Like @rm3195 said, we also define schema in JSON files.
Here's a sample of my resource config.
```terraform
resource "google_bigquery_table" "default" {
  count         = length(var.partitioned_tables)

  dataset_id    = var.dataset_id
  table_id      = replace(lookup(element(var.partitioned_tables, count.index), "table_name"), "-", "_")
  friendly_name = lookup(element(var.partitioned_tables, count.index), "friendly_name", null)
  description   = lookup(element(var.partitioned_tables, count.index), "description", null)

  time_partitioning {
    type  = "DAY"
    field = lookup(element(var.partitioned_tables, count.index), "partitioned_column")
  }

  labels = {
    env = "default"
  }

  schema = file("${path.root}/schemas/ddl/${var.dataset_id}-${replace(lookup(element(var.partitioned_tables, count.index), "table_name"), "-", "_")}.json")
}
```

**Comment by psvaiter** [comment]
The versions v3.54.0 and v3.55.0 changelogs mention a KNOWN ISSUE and a REVERT about this behavior of recreating a table. So it does not seem to be a coincidence in v3.56.0.

```
KNOWN ISSUES: New google_bigquery_table behaviour introduced in this version had unintended consequences, and may incorrectly flag tables for recreation. We expect to revert this for 3.55.0.
```

```
Reverted * bigquery: made incompatible changes to the google_bigquery_table.schema field to cause the resource to be recreated ([#8232](https://github.com/hashicorp/terraform-provider-google/pull/8232)) due to unintended interactions with a bug introduced in an earlier version of the resource.
```

**Comment by rm3195** [comment]
> @psvaiter yes, recreation is by design and expected when the schema is altered because it is a [JSON schema](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_table#schema) as a whole. Closing this issue then

Between version v3.32.0 and v3.52.0 (which we pinned now) ordering of fields is not taken into consideration (https://github.com/GoogleCloudPlatform/magic-modules/pull/3751). If order of the fields in JSON doesn't match JSON returned by API - terraform wouldn't show diff in versions <v3.53.0.
Also adding a field is accepted by GCP API and is not a destructive operation, it should be the same with Terraform.

**Comment by tpolekhin** [comment]
Hopefully im not beating a dead horse commenting on the closed issue, but I did some testing with various versions on the provider, and it behaves VERY differently each time.

So, our terraform code change was pretty simple: add 2 new columns to existing BigQuery table `SCHEDULE`
Nothing changes between runs - only provider version

# v3.52.0
```
Plan: 0 to add, 19 to change, 0 to destroy.
```
Mostly adds `+ mode = "NULLABLE"` to fields in bunch of tables, and adds 2 new fields in `SCHEDULE` table

# v3.53.0
```
Plan: 0 to add, 2 to change, 0 to destroy.
```
Adds 2 new fields to `SCHEDULE` table, and moves one field in another table in a different place (sorting?)

# v3.54.0
```
Plan: 1 to add, 1 to change, 1 to destroy.
```
Adds 2 new fields to `SCHEDULE` table, and moves one field in another table in a different place (sorting?) but now with table re-creation for some reason

# v3.55.0
```
Plan: 0 to add, 2 to change, 0 to destroy.
```
Adds 2 new fields to `SCHEDULE` table, and moves one field in another table in a different place (sorting?)
behaves exactly like v3.53.0

# v3.56.0
```
Plan: 1 to add, 0 to change, 1 to destroy.
```
Adds 2 new fields to `SCHEDULE` table but now wants to re-create table to do this

As @rm3195 pointed out in the BigQuery documentation, and how it was working for us for a long time now, adding new fields to existing table should not destroy and re-create table, effectively wiping all the data in it

Intended behaviour was implemented already in provider versions 3.53.0 and 3.55.0

@edwardmedia we would really appreciate if you could reconsider your decision, because we do a lot of table schema updates during our development process and loosing all the data each time is really frustrating

I hope this helps!
Thanks

**Comment by ScottSuarez** [comment]
Hi folks, definitely a bug!

https://github.com/GoogleCloudPlatform/magic-modules/pull/4606/files
Should fix this
